Implement Freeze/Unfreeze

tbg opened the pull request:
WIP but ready for a first look. If Reviewable shows you more than one revision, just
combine them all - there's been some rebasing.

See #6284. Implement this via a range-spanning RPC which is processed through
Raft and puts the Range in frozen mode. When frozen, all requests fail (except
freeze/unfreeze requests).

Added an admin endpoint which attempts to put the cluster in frozen mode (or
recover it from it) by querying ranges in the right order (so that the meta
addressing does not collapse before all ranges are frozen). This should work
with the exception of an ungossiped first range (left a TODO for this case;
doable).

When a process is starting, it will attempt to unfreeze any frozen Ranges for
which it has a Replica. This should always be able to recover the cluster (by
simply restarting a subset of nodes whose Replicas cover the key space).

Focus of discussion/open work:
- Discuss (preliminary) name `AdminSetFrozen`. It's not technically an
  AdminRequest but a normal ranged Raft write command (and it appears that
  that's the better choice). We could rename it to `SetFrozen` but then it's
  not clearly a command any more. We could also have a pair
  `Freeze`/`Unfreeze/Thaw` but that adds more boilerplate but isn't helpful.
  My opinion on having a range-admin command which then calls a second Raft
  write on the local range is similar: Much obfuscation, small upshot.
- Discuss semantics exposed maintenance endpoint. POST chosen since GET seems
  to innocuous for a feature like this.
- Write a whole lot of unit and acceptance tests. In particular
  - MinStartKey functionality
  - swift recovery of a multi-node cluster
  - the resulting upgrade process
    Have only tested this stuff locally on a single-node instance, so expect some
    more work to get everything in order.
- Should store nonce to avoid starting a node with a new version when it's
  not completely frozen. Better done now than later if we want it (because migration).

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6345)

<!-- Reviewable:end -->

bdarnell commented:
There's one big missing piece: SetFrozen returns when the leader replica has been frozen, but we need to wait for all followers to apply the freeze as well.  I'm not sure of the best way to do that.

---

Reviewed 20 of 21 files at r1.
Review status: 20 of 21 files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[roachpb/api.proto, line 214 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGNr7HxWkm_wajVEhme:-KGNr7Hy7J8mbd54HgmV:-952577454) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/roachpb/api.proto#L214)):*
I would call it `SetFrozen` instead of `AdminSetFrozen`.

---

*[roachpb/api.proto, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGNze17zlX1Ov8usTc6:-KGNze17zlX1Ov8usTc7:-901535775) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/roachpb/api.proto#L222)):*
It took me a long time to figure out why this is necessary, because I was looking in the wrong places for information about it. The comments from server/admin.go should be moved here.

Is it possible to get rid of this? I think it would be equivalent to make `Freeze` a noop if the request header's start key is greater than the range's start key. 

---

*[roachpb/api.proto, line 228 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGNsLGSlJFH-NXEiBmC:-KGNsLGSlJFH-NXEiBmD:-32095466) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/roachpb/api.proto#L228)):*
Needs a comment.

---

*[roachpb/errors.proto, line 80 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGNqzKu3r8wXA2ULY94:-KGNqzKu3r8wXA2ULY95:-2140696179) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/roachpb/errors.proto#L80)):*
Why does this contain a descriptor?

---

*[server/admin.go, line 772 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGO91-JYPNV4XV4ABQx:-KGO91-JYPNV4XV4ABQy:-170784633) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/server/admin.go#L772)):*
I don't think we need both of these since they're always in the first range.

---

*[server/admin.go, line 789 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGO8b6ckrroh0zXM80A:-KGO8b6ckrroh0zXM80B:-1983547632) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/server/admin.go#L789)):*
This could race with a split: a range which contained meta keys when the regular keys were frozen might be split by the time we come back to freeze the meta keys. 

We could solve this by having the freeze response return the smallest start key of any frozen range, and using that as the end key for the next freeze span.

---

*[storage/replica_command.go, line 1766 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGNyB8TrxrXsJPWMI7M:-KGNyB8TrxrXsJPWMI7N:71572944) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/replica_command.go#L1766)):*
Why does this `loadFrozenStatus` instead of using `r.mu.frozen`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KGO9X313ap2n7d4yLVv:874139834)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
The code looks good other than a lack of integration tests.

---

Reviewed 21 of 21 files at r1.
Review status: all files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[keys/constants.go, line 98 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOOvwUU05EuXVODu51:-KGOOvwVFsKdsEs-45E9:-1008677485) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/keys/constants.go#L98)):*
These seem to be sorted lexicographically, consider moving this up.

---

*[server/admin.go, line 770 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOS4W0rYw-Zis8m-gT:-KGOS4W1HwlbFqbMzl8p:1730991990) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/server/admin.go#L770)):*
Can we pull this into a static?

---

*[storage/replica.go, line 443 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOVHiU_izQcCsS6JTA:-KGOVHiVxQ73MjKP145w:1804765504) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/replica.go#L443)):*
`context.TODO()`?

---

*[storage/replica.go, line 448 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOVQq2HBl_hFFfoWEp:-KGOVQq2HBl_hFFfoWEq:1700997001) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/replica.go#L448)):*
Ditto.

---

*[storage/replica.go, line 1619 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOWX79ruAnm5OoRMLu:-KGOWX79ruAnm5OoRMLv:-2001606198) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/replica.go#L1619)):*
We only checked this if we got an error from `redirectOnOrAcquireLeaderLease`, right? Do we need to do this check unconditionally, or is it impossible to be frozen and not get an error back from `redirectOnOrAcquireLeaderLease`?

---

*[storage/replica_command.go, line 1793 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOXq-G0Veut86aFGwD:-KGOXq-HOGQr3fjTY38D:485661950) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/replica_command.go#L1793)):*
Do we want to do this if `setFrozenStatus()` fails?

---

*[storage/store.go, line 829 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOYagwGmhXMBCR-_Hk:-KGOYagwGmhXMBCR-_Hl:1804765504) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/store.go#L829)):*
`context.TODO()`?

---

*[storage/store.go, line 832 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOZTrEFdA1-uUrMB58:-KGOZTrEFdA1-uUrMB59:-420215606) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/store.go#L832)):*
Should this add `resp.Affected`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KGOZlxE0dGKs33GjOt3:-912800651)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Thanks for the reviews. Updated (you want to squash all the revisions since the last review; there's no interesting intermediate state).
I'm ready to start writing the tests unless there are major suggested changes.

@bdarnell: Good point. I added more (so far, untested) experiments to this end, please let me know if those sound copacetic:

* added MustVersion field to `AdminSetFrozen`. When a Freeze is applied, that is compared against the running version. On mismatch, explode.
* To avoid exploding in practice, added a "hook" at the end of a freeze which uses logic from the RaftLogQueue to determine which part of the log is "stale" (and injects new no-ops into Raft until that has happened). This needs some refactoring to look good, but I also want to know whether that's reasonable. Unfortunately, the `RaftStatus` fields aren't really that well-documented.

@nvanbenschoten: Yep, a lot of testing to be done. Will get to that as soon as @bdarnell confirms that we don't need to do something completely different.

---

Review status: 7 of 22 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---

*[keys/constants.go, line 98 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOOvwUU05EuXVODu51:-KGRDQUxFA7Lwa0gZHGT:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/keys/constants.go#L98)):*
Done.

---

*[roachpb/api.proto, line 214 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGNr7HxWkm_wajVEhme:-KGRE-Znqift-_Blj4tK:333451123) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/roachpb/api.proto#L214)):*
Done. (And renamed to ChangeFrozen to distinguish it from a getter).

---

*[roachpb/api.proto, line 222 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGNze17zlX1Ov8usTc6:-KGRHMRRjQZTMhiBOuTr:-1713432093) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/roachpb/api.proto#L222)):*
Good call, was able to remove the field (well, let's wait for the testing).
I left the comments where they were but added a comment around here that explains how this command works.

---

*[roachpb/api.proto, line 228 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGNsLGSlJFH-NXEiBmC:-KGRELtVFk-Xy6tg1zkQ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/roachpb/api.proto#L228)):*
Done.

---

*[roachpb/errors.proto, line 80 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGNqzKu3r8wXA2ULY94:-KGRH_3Cu6SsqJo3JDT_:1127272166) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/roachpb/errors.proto#L80)):*
In case we (or anyone really) has trouble with this and in particular in the case of ranges which refuse to unfreeze, this error will be received by all client traffic, so I wanted it to contain information about its source.

---

*[server/admin.go, line 770 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOS4W0rYw-Zis8m-gT:-KGRHxZAhv4SM12iu12n:900515574) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/server/admin.go#L770)):*
I could, but it's only used here and then I have to come up with a name that doesn't sound ridiculous but descriptive. If you feel strongly, I do take suggestions.

ps: this thing moved/changed.

---

*[server/admin.go, line 772 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGO91-JYPNV4XV4ABQx:-KGRIKkAMLKk4rPumFS8:-1869285803) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/server/admin.go#L772)):*
After the user ranges, you want to freeze meta2, so you need [some_key_before_all_meta2_ranges, Meta2KeyMax). You don't want to freeze the first range, so you can't use LocalMax. But you must use LocalMax in the last step to freeze the first range. Ergo you need another key in between and `Meta1KeyMax` looks fine. Missing something?

---

*[server/admin.go, line 789 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGO8b6ckrroh0zXM80A:-KGRJC3w4O1fBjdlw4i_:382891504) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/server/admin.go#L789)):*
We could also send the last request to all ranges. But I like your idea, 

---

*[storage/replica.go, line 443 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOVHiU_izQcCsS6JTA:-KGR_BjarnWphD7Oxbb0:-1348325098) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/replica.go#L443)):*
I've been using `context.Background()` in those cases which are far enough off the critical path and unlikely to actually be traced (the low-level plumbing for contexts is mostly to facilitate getting custom tracepoints in during debugging).
To my mind, `TODO()` should be reserved for those instances where we're really missing out on interesting info.

---

*[storage/replica.go, line 1619 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOWX79ruAnm5OoRMLu:-KGRaXt-xJ5f6jJ5idkD:154817021) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/replica.go#L1619)):*
Good catch, you're right. I moved the actual check down here.

---

*[storage/replica_command.go, line 1766 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGNyB8TrxrXsJPWMI7M:-KGRaj0z8WMySgIV0tlt:1024741018) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/replica_command.go#L1766)):*
Seems more idiomatic to read from the engine in a replica command than to grab the mutex, but the two will always be identical (mod bugs). Do you prefer it the other way?

---

*[storage/replica_command.go, line 1793 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOXq-G0Veut86aFGwD:-KGRb30y8EMOZrOC4IiW:-1757339338) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/replica_command.go#L1793)):*
It won't execute on error, so I saved myself the `err` dance.

---

*[storage/store.go, line 829 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOYagwGmhXMBCR-_Hk:-KGRbHFTH_BVUW1e31Em:1298477210) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/store.go#L829)):*
Yeah, this one should be a TODO (we should trace that guy).

---

*[storage/store.go, line 832 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGOZTrEFdA1-uUrMB58:-KGRc8sa6Vx9Xixdc7dh:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/store.go#L832)):*
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KGRCVnGXgrc2d0EfWbz:919248562)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@cuongdo could you look over the endpoint definition (`admin.proto`)? Just want to make sure that if there's a plan for laying out the API, I'm following it.

---

Review status: 7 of 22 files reviewed at latest revision, 15 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KGRicRwZV8tVd3F_w23:-1092886720)*
<!-- Sent from Reviewable.io -->

cuongdo commented:
@tschottdorf: admin proto stuff looks good, with a small nit that you can take or leave

---

Review status: 7 of 22 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[server/admin.proto, line 249 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGS85j-HQGQz-MZ5CaL:-KGS85j0b6Znn5NYp7uz:-1325752650) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/server/admin.proto#L249)):*
I'd prefer that this name be a little more self-documenting, like `ranges_affected`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KGS8c_c4KkuwpRQ6xm5:1418477755)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Thanks @cuongdo.

---

Review status: 7 of 22 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[server/admin.proto, line 249 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGS85j-HQGQz-MZ5CaL:-KGS9-tAzdV127vOoAd1:-1363907954) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/server/admin.proto#L249)):*
That's very reasonable, changed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KGS8zgTKf-JYrhMr4wi:-367415461)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 10 of 15 files at r2, 3 of 5 files at r3, 3 of 3 files at r4.
Review status: 21 of 22 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[server/admin.proto, line 249 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGS85j-HQGQz-MZ5CaL:-KGTCM7Hy5GufWQJqZT9:-1089288625) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/server/admin.proto#L249)):*
Use the same name in `roachpb.ChangeFrozenResponse`

---

*[storage/replica.go, line 956 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTHdNy18yyHHntZrGa:-KGTHdNy18yyHHntZrGb:306624523) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica.go#L956)):*
This is the wrong place for this. `endCmds` runs on all replicas, but `getTruncatableIndexes` only works on the raft leader (it returns zeros elsewhere so this will just block for the duration on followers). 

Furthermore, the node that proposed the command (which is the only one where this blocking matters) may have timed out and stopped waiting (returning an error upstream). I think we need to do this separately from blocking in the command itself, but I'm not sure of the right way to do it. 

---

*[storage/replica_command.go, line 1766 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGNyB8TrxrXsJPWMI7M:-KGTDizut7nqNohjxcn_:-1639275771) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/replica_command.go#L1766)):*
I don't feel strongly but I think we generally go the other way (most often with `r.Desc()`).

---

*[storage/replica_command.go, line 1749 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTEHEJo-us4wKIkMKp:-KGTEHEJo-us4wKIkMKq:1303361962) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica_command.go#L1749)):*
The version check means that this is not idempotent (the actual idempotence comes from the fact that the second freeze never even makes it this far). 

---

*[storage/replica_command.go, line 1771 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTE__FJH7iFCIl5A0d:-KGTE__FJH7iFCIl5A0e:1225188156) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica_command.go#L1771)):*
This scares me. Requiring an exact match seems unnecessarily strict.  Marking one node as corrupt is OK, but what if your one node with a different version from the others happens to be the one where you initiate the freeze? The entire rest of the cluster would be marked as corrupt with no good way to recover (if `replicaCorruptionError` were finished. Currently it doesn't do anything, so the "corrupt" state will be lost on restart and the node will restart as if nothing happened).

---

*[storage/store.go, line 818 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTDAiRUTbMpha_0Oyu:-KGTDAiRUTbMpha_0Oyv:-390961406) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/store.go#L818)):*
Check the return value of `RunLimitedAsyncTask` or the `WaitGroup` may never stop waiting.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
@bdarnell I'll take your review to imply that this is basically looking like it's going to be a mergable PR and begin working on tests/... in addition to bouldering out the rest of your comments.

---

Review status: 21 of 22 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 956 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTHdNy18yyHHntZrGa:-KGTTQGi_ZWm9nQe2o0k:-515938022) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica.go#L956)):*
Yeah, I'll refactor this to the right place - but this would generally work/seem like the most reasonable idea?
I don't think it matters if the client stopped waiting. If you get an error during a freeze operation, you'll always end up with some ranges frozen and others unfrozen. Try again and you'll be ok. Or not, in which case you probably want to unfreeze instead and file a bug.

---

*[storage/replica_command.go, line 1771 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTE__FJH7iFCIl5A0d:-KGTTsIk4Hyb5bdVzEtP:1899133458) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica_command.go#L1771)):*
Ok, what do you want to happen? I can move the tag below the idempotence check (in fact I meant to only check this when we actually freezing the Replica). I think if you mess up a migration, there will always be a way to break the cluster, and we can throw up or pretend it didn't happen. We could also check the version before propose time, but that means looking into every batch in the critical path (maybe that's ok because we can guard it with `len(ba.Requests)==1`). WDYT?

Btw, the second command does make it that far. Freeze/Unfreeze always go through to avoid giving the callers errors which prevent them from finishing or undoing an operation on a partially frozen cluster.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KGTUuExhhCkx0iRCiAm:-1159797211)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Yes, everything looks good except for how we wait for it to be applied everywhere.

---

Review status: 21 of 22 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 956 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTHdNy18yyHHntZrGa:-KGTWv2L-rDESIQJepBn:-842663842) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica.go#L956)):*
What are you thinking is the right place for this?

The problem is that if the node that the client is waiting on is not the raft leader, we don't have any visibility into client log positions. And actually that's not even the right thing: the raft leader knows the log index the followers have *acknowledged*, not *applied*.

There's not anything that communicates the applied index of every replica, so there's not an existing mechanism that could be used to watch this. I don't know what the right solution will look like.

---

*[storage/replica_command.go, line 1771 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTE__FJH7iFCIl5A0d:-KGTYchE8DC1V0H8PCUC:881623164) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica_command.go#L1771)):*
I'd lean towards "pretend it didn't happen" for now. After we have proper `replicaCorruptionErrors` then I'd say we could put this back in and have the client first poll every node (or use gossip?) to make sure that everyone is on the same version before starting the process.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KGTWk1zmHKjOvojNDEi:822327984)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Ok. I'll work on all the tests now and, for the time being, pretend that the wait-for-raft-acks mechanism is actually sufficient (clearly isn't). I think waiting for a command to be applied on all followers should be tackled separately as a follow-up since there are various ways of going about it.

---

Review status: 21 of 22 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[server/admin.proto, line 249 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGS85j-HQGQz-MZ5CaL:-KGkRJBwTrK6_tbSxko9:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/server/admin.proto#L249)):*
Done.

---

*[storage/replica.go, line 956 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTHdNy18yyHHntZrGa:-KGkRav0FNq6xaJe_wBc:1440575668) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica.go#L956)):*
> What are you thinking is the right place for this?

This is the right "place", just buried in a weird location. I want the command to exit the command queue but then not acknowledge upstream of `*Replica` until it can make some guarantees about what's been applied (this check here doesn't go far enough, but I hope that it can serve as a substitute for a new mechanism we need here).
If some upstream mechanism times out, that's ok. The command can retry.

---

*[storage/replica_command.go, line 1766 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGNyB8TrxrXsJPWMI7M:-KGkWmBK0rOeUcKjtlYx:1577770251) ([raw file](https://github.com/cockroachdb/cockroach/blob/6617ca9ac3c44df216f0ecda3fef7f2ef26f10a1/storage/replica_command.go#L1766)):*
Ok. I'll leave it for now (no strong preference here as well, but it is an explicit lock acquisition and somehow the current code appeals to me more).

---

*[storage/replica_command.go, line 1749 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTEHEJo-us4wKIkMKp:-KGkSPk_VUMT7Wcar4R6:-738490820) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica_command.go#L1749)):*
As mentioned below, all freeze/unfreeze commands should make it all the way through.

---

*[storage/replica_command.go, line 1771 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTE__FJH7iFCIl5A0d:-KGkSXuJlTJQTgBi1B5Z:1108382225) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica_command.go#L1771)):*
I don't fully understand the reasoning - we are doing all of this to avoid unintentionally mixing versions during migrations. Seems awfully hard to guarantee or even test for if we don't check it?
Yes, there is some care we should take to avoid that someone can immediately wedge their cluster by proposing a freeze from the wrong version, but with the above ideas that would only happen if you're running a mixed cluster. We *want* to make that fail (not irrecoverably, but still), no?
I understand that you're worried about folks accidentally blowing out their cluster, but I'd rather be strict now.

---

*[storage/store.go, line 818 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTDAiRUTbMpha_0Oyu:-KGkXRi3eVN-Peq7MlvO:926693478) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/store.go#L818)):*
Good point, done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KGkXTicBlruLmm7xKcG:-663904510)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 21 of 22 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 956 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTHdNy18yyHHntZrGa:-KGkzc92a1bQE6xuI-tP:1064590576) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica.go#L956)):*
OK. I was wrong earlier when I said `endCmds` ran on all replicas; it's lease-holder-only, so that's fine. I think it might be a little cleaner to do this in `addWriteCmd` instead of `endCmds`, but it doesn't make much difference.

---

*[storage/replica_command.go, line 1771 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTE__FJH7iFCIl5A0d:-KGl-OPBd9kryZlHCV00:-630405302) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica_command.go#L1771)):*
> I don't fully understand the reasoning - we are doing all of this to avoid unintentionally mixing versions during migrations.

Most of the work we're doing in this change is to ensure that versions are not mixed in a cluster whose operator is following all the steps correctly. This check, though, is different because it's trying to make the process safer in the presence of certain kinds of mistakes. That's a laudable goal, but I don't think this is the right place for it. Our primary safeguard against mixed clusters should be before the process even starts. Once it gets to this point, a `replicaCorruptionError` has too much potential for damage and data loss (a panic/crash is safer in this context since you should be able to restart with the correct (older) version, although that's not guaranteed to be safe either)

>  if you're running a mixed cluster. We want to make that fail (not irrecoverably, but still), no?

It's safest if a cluster is all on one version, but it's reasonable to run a (slightly) different version on some nodes (for example, running one node with jemalloc profiling with the standard build on the rest). I think it's fine to refuse to freeze mixed clusters if we catch it early enough.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Wrote the basic tests and, not entirely unexpectedly, once you kill a node and try to freeze the cluster it wedges itself in a bad state (have a hard time telling what exactly is going wrong, but it's not the `endCmds` trigger's fault - without it, same behavior). Will investigate that more, but the code that is now pushed should be reviewed.

---

Review status: 21 of 22 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 956 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTHdNy18yyHHntZrGa:-KGlUkvFD1qe9qZS80VP:417681260) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica.go#L956)):*
Shuffled things around a little bit (in particular to keep `endCmds` small). Still worthy of refactoring imo, but there will be a lot of refactoring coming in `Replica` anyway, so I might leave it where it is for this PR.

---

*[storage/replica_command.go, line 1771 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTE__FJH7iFCIl5A0d:-KGlUWcJqf6RYDBz-hOo:1710491866) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica_command.go#L1771)):*
Ok, I changed the returned error to a "regular" error (i.e. if something goes wrong some replicas will log the error, but nobody will care) and left a comment. At the same time, I did introduce a pre-proposal safeguard - can't hurt.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KGlUwiSYfZEGkhvh3Oo:523464598)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
BTW, your comment on not using `info.Tag` for versioning makes sense. What is the right thing to use? looks like it's time to introduce a Raft command version?

---

Review status: 8 of 29 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KGlWlnaakhtSq1ZuSQs:63780079)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Yeah, I think the version check should be based on an explicit `raftCommandVersion` (or `freezeVersion`? Will we use freezing for non-raft upgrades or will we be able to just do a normal drain?) instead of a build tag.

---

Reviewed 20 of 21 files at r5, 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[acceptance/update_test.go, line 38 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGlkWpulz-3QYrYtg-r:-KGlkWpwKIyRQmqxyIUQ:138491763) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fc67b9e53f888f66943bb95646e928a9b72aa3/acceptance/update_test.go#L38)):*
Call this `postFreeze`. Can't you use `postJSON` for most of this?

---

*[acceptance/update_test.go, line 66 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGll3KlgobLoOZFDmJX:-KGll3Km2kUEs4G8gs2x:1515661685) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fc67b9e53f888f66943bb95646e928a9b72aa3/acceptance/update_test.go#L66)):*
Use `server.ExpectedInitialRangeCount()`?

---

*[storage/replica_command.go, line 1771 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTE__FJH7iFCIl5A0d:-KGlllBQSSULJh6q-XQP:1417477088) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica_command.go#L1771)):*
Returning a regular error will cause the replicas to diverge as some have frozen and some have not. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KGlmG8HVdhqI-0v5Voe:995125890)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
I *think* for simplicity we should just have all stop-the-world upgrades employ freezing (easy enough with a "shutdown for upgrade" tool) to avoid burdening users with knowing what update is what type (except online/offline).

PTAL - The tests work reliably now, at least locally. Do we want to figure the versioning out in this open PR or separately?

---

Review status: 23 of 32 files reviewed at latest revision, 8 unresolved discussions, some commit checks pending.

---

*[acceptance/update_test.go, line 38 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGlkWpulz-3QYrYtg-r:-KGmfWtS4bWNziTXbKoT:-449546607) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fc67b9e53f888f66943bb95646e928a9b72aa3/acceptance/update_test.go#L38)):*
Yes and done. Fixed a buglet in `postJSON` (and the old code here) that didn't bother to check the HTTP error code. Took me longer than it should have to find it.

---

*[acceptance/update_test.go, line 66 \[r6\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGll3KlgobLoOZFDmJX:-KGmfqVmwjZRT2UVMcR8:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/68fc67b9e53f888f66943bb95646e928a9b72aa3/acceptance/update_test.go#L66)):*
Done.

---

*[storage/replica_command.go, line 1771 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGTE__FJH7iFCIl5A0d:-KGmiYU9rC67etEySWka:-1363490947) ([raw file](https://github.com/cockroachdb/cockroach/blob/856a26bd60aac46e515a7f8ef839b3bf0ea6d66d/storage/replica_command.go#L1771)):*
Whoops, of course. Fixed. I'm still weirded out by not exploding here, but I suppose we'll reintroduce this at some point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KGmil1lwqjk_zeyuJc9:1314308779)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

I'm fine with leaving everything related to versioning out of this PR; we can leave version paranoia and checks that all replicas have been frozen for a future PR. They may even use similar mechanisms: poll all nodes beforehand for their version, then poll all nodes afterwards for their number of frozen vs unfrozen ranges. 

---

Reviewed 9 of 9 files at r7.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 976 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGnjHHPDDpxny1n-lVy:-KGnjHHQouMF6TB-8akK:-2047114680) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dff14c41a31675d7ab070f3b04d177917dca0f5/storage/replica.go#L976)):*
The comparison `waitForIndex >= oldestInUse` looks right to me.

---

*[storage/replica.go, line 981 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGnjBXLFaGcv4-Hkc9m:-KGnjBXMuwbb0sz29IOz:-1292574031) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dff14c41a31675d7ab070f3b04d177917dca0f5/storage/replica.go#L981)):*
This shouldn't be necessary; raft should advance the other replicas on its own. 

---

*[storage/replica.go, line 1009 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGnjfxseRRJG8YSOjbg:-KGnjfxseRRJG8YSOjbh:-358237453) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dff14c41a31675d7ab070f3b04d177917dca0f5/storage/replica.go#L1009)):*
Comment that this `appliedIndex` is not necessarily the index of the command we're ending here (another thread could have applied another command since then), but it's guaranteed to be greater than or equal to the index of our command, which is good enough.

---

*[storage/replica.go, line 1348 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGnkCAmKJWtZyvr14DF:-KGnkCAn-AfAKAg7Vqwj:-2001060654) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dff14c41a31675d7ab070f3b04d177917dca0f5/storage/replica.go#L1348)):*
I don't like adding this special case at proposal time; I know this is unavoidably a special command in many ways but this doesn't have to be one of them (this code also only runs on the leader, so it's not sufficient). I was thinking that a pre-check of versions would happen on the admin/client side before the ChangeFrozen command even starts, by reading from every node.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KGnlprlRo1q3-t8WLro:2144934862)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/replica.go, line 976 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGnjHHPDDpxny1n-lVy:-KGwbSb8M8n8YjySxEKs:842344226) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dff14c41a31675d7ab070f3b04d177917dca0f5/storage/replica.go#L976)):*
But in the absence of other activity, `raftStatus.Applied` would be the appliedIndex of the freeze itself, and so we'd always have equality here, no? Fairly sure this failed without applying random stuff below.

Edit: yep, fails without this.
```
--- FAIL: TestStoreChangeFrozen (3.10s)
	store_test.go:1908: timed out waiting for Range range_id:1 start_key:"" end_key:"\377\377" replicas:<node_id:1 store_id:1 replica_id:1 > next_replica_id:2  to freeze: storage/replica.go:993: waiting for index 13 to be unused (still using 13)
```

So, off-by-one maybe in `getTruncatableIndexes` or do we really need to fake activity?

---

*[storage/replica.go, line 981 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGnjBXLFaGcv4-Hkc9m:-KGwdeSjuJBHxCN-yLzZ:-1582418045) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dff14c41a31675d7ab070f3b04d177917dca0f5/storage/replica.go#L981)):*
See above.

---

*[storage/replica.go, line 1009 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGnjfxseRRJG8YSOjbg:-KGwduS08qMGRY-MmM_w:1437694486) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dff14c41a31675d7ab070f3b04d177917dca0f5/storage/replica.go#L1009)):*
The trigger actually no-ops in that case (checks `RangesAffected`). I hoisted that code here instead.

---

*[storage/replica.go, line 1348 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGnkCAmKJWtZyvr14DF:-KGwfdq-EnecuooPSdRz:590083684) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dff14c41a31675d7ab070f3b04d177917dca0f5/storage/replica.go#L1348)):*
Ok, I removed this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 23 of 32 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 976 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGnjHHPDDpxny1n-lVy:-KH-BlptqOcvNip9GhO9:-805260058) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dff14c41a31675d7ab070f3b04d177917dca0f5/storage/replica.go#L976)):*
I was confused above; the comparison should be `waitForIndex > oldestInUse` (and s/oldestInUse/oldestAcknowledged/ for clarity). We're waiting for each replica to acknowledge `waitForIndex`. 

---

*[storage/replica.go, line 1002 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KH-CousS27lpi295JBd:-KH-CousS27lpi295JBe:829846586) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e32bf62d8b987ccd74accc89ca0c3c2d982fb65/storage/replica.go#L1002)):*
The deferred function accesses `r.mu.appliedIndex`, so doesn't it need to run inside the lock? 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 17 of 32 files reviewed at latest revision, 8 unresolved discussions.

---

*[storage/replica.go, line 976 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGnjHHPDDpxny1n-lVy:-KH-EpAxk-H4zmgeeoxJ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dff14c41a31675d7ab070f3b04d177917dca0f5/storage/replica.go#L976)):*
Done.

---

*[storage/replica.go, line 1002 \[r8\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KH-CousS27lpi295JBd:-KH-DklNOWslBcTMvK5f:-1538918300) ([raw file](https://github.com/cockroachdb/cockroach/blob/2e32bf62d8b987ccd74accc89ca0c3c2d982fb65/storage/replica.go#L1002)):*
You're right, lost during a reshuffle - the index needs to passed into the closure. Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 5 of 9 files at r8, 4 of 10 files at r9.
Review status: 26 of 32 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 981 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGnjBXLFaGcv4-Hkc9m:-KH-J-1ipKWKkItdSWuy:-1902006687) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dff14c41a31675d7ab070f3b04d177917dca0f5/storage/replica.go#L981)):*
Can't this be removed now that you've changed the comparison above?

---

*[storage/replica.go, line 1009 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGnjfxseRRJG8YSOjbg:-KH-JOfXG-NfNUkhoN2b:469046955) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dff14c41a31675d7ab070f3b04d177917dca0f5/storage/replica.go#L1009)):*
How does this response relate to my comment? My point was that at this point we are "ending" a freeze command, but we can't know at what index it was committed (but that's ok, because r.mu.appliedIndex gives us an upper bound).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345)*
<!-- Sent from Reviewable.io -->

tbg committed:
Implement Freeze RPC

See #6284. Implement this via a range-spanning RPC which is processed through
Raft and puts the Range in frozen mode. When frozen, all requests fail (except
freeze/unfreeze requests).

Added an admin endpoint which attempts to put the cluster in frozen mode (or
recover it from it) by querying ranges in the right order (so that the meta
addressing does not collapse before all ranges are frozen). This should work
with the exception of an ungossiped first range (left a TODO for this case;
doable).

ChangeFrozen contains a version string (build tag) but it is currently only
used to emit a warning (where in reality, the replica could be corrupted).
in the case in which a Replica which is about to freeze realizes that the
freeze was proposed by a different version.

A ChangeFrozen request which successfully froze a Replica will *attempt* to
ensure that the command was in fact applied by *all* Replicas before returning
to the client by inspecting Raft log forward status. This in turn means that
attempting to ChangeFrozen on a degraded-service cluster is ill-fated (but that
should be obviously not attempted in practice). The code doing that does
currently not guarantee it; this will be tackled separately.

When a Store is starting, it will attempt to unfreeze any frozen Ranges for
which it has a Replica. This should always be able to recover the cluster (by
simply restarting a subset of nodes whose Replicas cover the key space).

tschottdorf commented:
Review status: 26 of 32 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/replica.go, line 981 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGnjBXLFaGcv4-Hkc9m:-KH-LEXuoSDNFJiGeAYm:-1306430188) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dff14c41a31675d7ab070f3b04d177917dca0f5/storage/replica.go#L981)):*
Yep. Apparently I'm still asleep. Fixed.

---

*[storage/replica.go, line 1009 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-KGnjfxseRRJG8YSOjbg:-KH-KTmKFZ1wkOmW9YA2:1321424661) ([raw file](https://github.com/cockroachdb/cockroach/blob/1dff14c41a31675d7ab070f3b04d177917dca0f5/storage/replica.go#L1009)):*
Oh, I see. I missed that even at this point in the code the index could have increased - I thought you were referring to the fact that the old code was getting the index from the trigger. Added the comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 25 of 32 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6345#-:-KH-PRpQQ3dYjqzoBHRT:1417088721)*
<!-- Sent from Reviewable.io -->
